lib/rust/parser/src/syntax/tree.rs

```diff
@@ -194,11 +230,140 @@ impl<'s> span::Builder<'s> for Error {
+    expressions: impl IntoIterator<Item = Line<'s>>,
+) -> Tree<'s> {
+    use crate::expression_to_statement;
+    let expressions = expressions.into_iter();
+    let statements = expressions.map(|line| line.map_expression(expression_to_statement));
+    let statements = statements.collect();
+    Tree::body_block(block_start_newline, statements)
+}
+
+
+// === OperatorBlock ===
+
+/// The content of a line in an operator block.
+#[derive(Debug, Clone, PartialEq, Eq, Visitor, Reflect, Serialize, Deserialize)]
+pub struct OperatorBlockExpression<'s> {
+    /// The operator at the beginning of the line.
+    pub operator:   OperatorOrError<'s>,
+    /// The rest of the expression.
+    pub expression: Tree<'s>,
+}
+
+/// Interpret the given expression as an `OperatorBlockExpression`, if it fits the correct pattern.
+fn to_operator_block_expression(
+    expression_: Tree<'_>,
+) -> Result<OperatorBlockExpression<'_>, Tree<'_>> {
+    let tree_ = match &*expression_.variant {
+        Variant::OprSectionBoundary(OprSectionBoundary { ast }) => ast,
+        _ => return Err(expression_),
+    };
+    if let Variant::OprApp(OprApp { lhs: None, opr, rhs: Some(expression) }) = &*tree_.variant {
+        if expression.span.left_offset.visible.width_in_spaces < 1 {
+            return Err(expression_);
+        }
+        let mut operator = opr.clone();
+        first_operator_mut(&mut operator).left_offset = expression_.span.left_offset;
+        let expression = expression.clone();
+        Ok(OperatorBlockExpression { operator, expression })
+    } else {
+        Err(expression_)
+    }
+}
+
+impl<'s> span::Builder<'s> for OperatorBlockExpression<'s> {
+    fn add_to_span(&mut self, span: Span<'s>) -> Span<'s> {
+        span.add(&mut self.operator).add(&mut self.expression)
+    }
+}
+
+/// A line in an operator block.
+#[derive(Debug, Clone, PartialEq, Eq, Visitor, Reflect, Serialize, Deserialize)]
+pub struct OperatorBlockLine<'s> {
+    /// The operator-expression, if any.
+    pub expression: Option<OperatorBlockExpression<'s>>,
+    /// The end-of-line token.
+    pub newline:    token::Newline<'s>,
+}
+
+impl<'s> From<token::Newline<'s>> for OperatorBlockLine<'s> {
+    fn from(newline: token::Newline<'s>) -> Self {
+        Self { expression: None, newline }
+    }
+}
+
+impl<'s> span::Builder<'s> for OperatorBlockLine<'s> {
+    fn add_to_span(&mut self, span: Span<'s>) -> Span<'s> {
+        span.add(&mut self.expression).add(&mut self.newline)
+    }
+}
+
+
 // === OprApp ===
 
 /// Operator or [`MultipleOperatorError`].
 pub type OperatorOrError<'s> = Result<token::Operator<'s>, MultipleOperatorError<'s>>;
 
+/// Return a reference to the first operator, whether that is the only operator, or the first of the
+/// `MultipleOperatorError` sequence.
+pub fn first_operator<'s, 'a>(opr: &'a OperatorOrError<'s>) -> &'a token::Operator<'s> {
+    match opr {
+        Ok(opr) => opr,
+        Err(oprs) => oprs.operators.first(),
+    }
+}
```

Comment:
why is it not a method? If you want to keep it as a global function, it needs better name, because now it is `tree:: first_operator` and its confusing. The same applies below.